File: unique-binary-search-trees.js
/**
 * @param {number} n
 * @return {number}
 */
var numTrees = function(n) {
  var cache = [1, 1, 2];
  var numOfTrees = 0;
  var numTreesHelper = function(n) {
    if(!cache[n]) {
      for(var i = 0; i < n; i++) {
        numOfTrees += numTreesHelper(i) * numTreesHelper(n - 1 - i);
      }
      cache[n] = numOfTrees;
    }
    return cache[n];
  };
  return numTreesHelper(n);
};

File: unique-binary-search-trees.c
int numTrees(int n) {
  static int cache[1024] = {1, 1, 2, 5};
  int numOfTrees = 0;
  if(!cache[n]) {
    for(int i = 0; i < n; i++) {
      numOfTrees += numTrees(i) * numTrees(n - 1 - i);
      cache[n] = numOfTrees;
    }
  }
  return cache[n];
}
